### PR TITLE
Add a solution to the WhatsApp Cloud payload character limitation

### LINKED ISSUE #1604: [WhatsApp Cloud] Add a solution to the WhatsApp Cloud payload character limitation
Cette mise à jour apporte une solution pour la limitation de caractère des payloads

### PR DESCRIPTION
**Pull request by SNCFamiraBenkaci**
resolve #1604

### CODE REVIEW

File: bot/connector-whatsapp-cloud/src/main/kotlin/database/repository/PayloadWhatsAppCloudMongoDAO.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import com.mongodb.client.model.CreateCollectionOptions
+import com.mongodb.client.model.IndexOptions
+import com.mongodb.client.model.ReplaceOptions
+import mu.KotlinLogging
+import org.litote.kmongo.*
+import java.util.concurrent.TimeUnit
+
+object PayloadWhatsAppCloudMongoDAO: PayloadWhatsAppCloudDAO {
+
+    private const val COLLECTION_NAME = "sncv_whatsapp_payload"
```
Comment (vsct-jburet) [inline bot/connector-whatsapp-cloud/src/main/kotlin/database/repository/PayloadWhatsAppCloudMongoDAO.kt:35]: I would prefer to write 

private val collectionName = property("tock_whatsapp_payload", "whatsapp_payload")

File: bot/connector-whatsapp-cloud/src/main/kotlin/model/send/message/content/WhatsAppCloudBotInteractive.kt [hunk 2]
```diff
@@ -94,7 +94,7 @@ data class WhatsAppCloudBotActionButton(
 
 data class WhatsAppCloudBotActionButtonReply(
     val title: String,
-    val id: String,
+    var id: String,
```
Comment (vsct-jburet) [inline bot/connector-whatsapp-cloud/src/main/kotlin/model/send/message/content/WhatsAppCloudBotInteractive.kt:97]: use val and copy() when you need to change it

File: bot/connector-whatsapp-cloud/src/main/kotlin/model/send/message/content/WhatsAppCloudBotTemplate.kt [hunk 3]
```diff
@@ -62,7 +62,7 @@ sealed class Component {
 
 
 data class TextParameter(val type: ParameterType, val text: String)
-data class PayloadParameter(val type: ParameterType, val payload: String?, val text: String)
+data class PayloadParameter(val type: ParameterType, var payload: String?, val text: String)
```
Comment (vsct-jburet) [inline bot/connector-whatsapp-cloud/src/main/kotlin/model/send/message/content/WhatsAppCloudBotTemplate.kt:65]: use val

File: bot/connector-whatsapp-cloud/src/main/kotlin/model/webhook/message/content/ButtonContent.kt [hunk 4]
```diff
@@ -17,6 +17,6 @@
 package ai.tock.bot.connector.whatsapp.cloud.model.webhook.message.content
 
 data class ButtonContent(
-        val payload: String,
+        var payload: String,
```
Comment (vsct-jburet) [inline bot/connector-whatsapp-cloud/src/main/kotlin/model/webhook/message/content/ButtonContent.kt:20]: use val

File: bot/connector-whatsapp-cloud/src/main/kotlin/model/webhook/message/content/InteractiveContent.kt [hunk 5]
```diff
@@ -27,7 +27,7 @@ data class InteractiveContent(
 )
 
 data class ButtonReply(
-        val id: String,
+        var id: String,
```
Comment (vsct-jburet) [inline bot/connector-whatsapp-cloud/src/main/kotlin/model/webhook/message/content/InteractiveContent.kt:30]: use val

File: bot/connector-whatsapp-cloud/pom.xml [hunk 6]
```diff
@@ -79,6 +79,14 @@
...
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.litote.kmongo</groupId>
+            <artifactId>kmongo</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.litote.kmongo</groupId>
+            <artifactId>kmongo-async</artifactId>
+        </dependency>
```
Comment (vsct-jburet) [inline bot/connector-whatsapp-cloud/pom.xml:89]: remove kmongo-async - it should work with kmongo only

**Comment by SNCFamiraBenkaci** [comment]
> #1604

File: bot/connector-whatsapp-cloud/pom.xml (hunk 6 above)
Comment (Fabilin) [reply bot/connector-whatsapp-cloud/pom.xml:89]: Just wondering, what's the reason ?

File: bot/connector-whatsapp-cloud/src/main/kotlin/Ioc.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+package ai.tock.bot.connector.whatsapp.cloud
+
+import ai.tock.bot.connector.whatsapp.cloud.database.repository.PayloadWhatsAppCloudDAO
+import ai.tock.bot.connector.whatsapp.cloud.database.repository.PayloadWhatsAppCloudMongoDAO
+import ai.tock.shared.service.BotAdditionalModulesService
+import com.github.salomonbrys.kodein.Kodein
+import com.github.salomonbrys.kodein.bind
+import com.github.salomonbrys.kodein.singleton
+
+val whastappCloudConnectorModule = Kodein.Module {
```
Comment (vsct-jburet) [inline bot/connector-whatsapp-cloud/src/main/kotlin/Ioc.kt:26]: typo **whats**AppCloudConnectorModule

File: bot/connector-whatsapp-cloud/src/main/kotlin/database/repository/PayloadWhatsAppCloudMongoDAO.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import com.mongodb.client.model.CreateCollectionOptions
+import com.mongodb.client.model.IndexOptions
+import com.mongodb.client.model.ReplaceOptions
+import mu.KotlinLogging
+import org.litote.kmongo.*
+import java.util.concurrent.TimeUnit
+
+object PayloadWhatsAppCloudMongoDAO: PayloadWhatsAppCloudDAO {
+
+    private val COLLECTION_NAME = property("tock_whatsapp_payload", "whatsapp_payload")
```
[the remark this answers is not in the document]
Comment (vsct-jburet) [reply bot/connector-whatsapp-cloud/src/main/kotlin/database/repository/PayloadWhatsAppCloudMongoDAO.kt:38]: this is not a const so private `val collectionName = `

File: bot/connector-whatsapp-cloud/src/main/kotlin/database/repository/PayloadWhatsAppCloudMongoDAO.kt [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                PayloadWhatsAppCloud::payloadExpireDate,
+                indexOptions = IndexOptions().expireAfter(1L, TimeUnit.DAYS)
+            )
+        } catch (e: Exception) {
+            logger.error(e)
+        }
+    }
+
+    fun isUUID(uuid: String): Boolean {
+        val uuidRegex = Regex("[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}")
```
Comment (vsct-jburet) [inline bot/connector-whatsapp-cloud/src/main/kotlin/database/repository/PayloadWhatsAppCloudMongoDAO.kt:91]: This should be a val at class level in order to avoir calculate the regexp each time

**Review by vsct-jburet — APPROVED** [review]

**Pull request merged**